Species CSV

nspool opened the pull request:
This PR includes all of the work done in https://github.com/aodn/nrmn-application/pull/107 as well as the additional species attributes, site specific letter code and tests.

Stack: http://nrmn-species-csv.dev.aodn.org.au/data/extract

utas-tajona commented on api/src/main/java/au/org/aodn/nrmn/restapi/model/db/SpeciesWithAttributes.java:
```diff
@@ -0,0 +1,47 @@
+@Data
+@Table(name = "ui_species_attributes")
+public class SpeciesWithAttributes {
+
```
are you using this Entity? it has been removed. validatition is using the interface: UiSpeciesAttributes

nspool replied:
Yes, it's used by SpeciesWithAttributesRepository. This was added in the earlier PR this one was based of. See https://github.com/aodn/nrmn-application/pull/107.

jonescc replied:
This doesn't look like its being used - can we remove it?  We should consolidate species with attributes methods in the process.

utas-tajona commented on api/src/main/java/au/org/aodn/nrmn/restapi/repository/projections/SpeciesWithAttributesCsvRow.java:
```diff
@@ -0,0 +1,23 @@
+public class SpeciesWithAttributesCsvRow {
+
+    private String letterCode;
+
```
i think you can implement  UiSpeciesAttributes and remove all the duplicate fields

nspool replied:
Looked into this and it could be done, but not without substantial changes to one or the other usage. Also the CsvRow has the lettercode which I think makes it clearer to keep seperate.

utas-tajona commented on api/src/main/java/au/org/aodn/nrmn/restapi/repository/SpeciesWithAttributesRepository.java:
```diff
@@ -0,0 +1,27 @@
+
+@Repository
+public interface SpeciesWithAttributesRepository
+        extends JpaRepository<SpeciesWithAttributes, Integer>, JpaSpecificationExecutor<SpeciesWithAttributes> {
```
do you need this ? this repo was removed longtime ago

nspool replied:
It's still used in getSpeciesForTemplate.

utas-tajona replied:
ui_species_attribute is pulled from ObservationRepository as a query 
https://github.com/aodn/nrmn-application/blob/master/api/src/main/java/au/org/aodn/nrmn/restapi/repository/ObservationRepository.java

utas-tajona commented on api/src/main/java/au/org/aodn/nrmn/restapi/service/TemplateService.java:
```diff
@@ -53,19 +74,56 @@ public void writeZip(OutputStream outputStream,
+
+        List<Integer> siteIds = sites.stream().map(s -> s.getSiteId()).collect(Collectors.toList());
+        List<LetterCodeMapping> letterCodeMappings = letterCodeRespository.findByCriteria(siteIds);
+        HashMap<Integer, String> letterCodeMap = new HashMap<Integer, String>();
+        letterCodeMappings.forEach(m -> letterCodeMap.put(m.getObservableItemId(), m.getLetterCode()));
```
This to line can be inline to:
 ``` 
HashMap<Integer, String>  letterCodeMap = letterCodeMappings.stream().collect(Collectors.toMap(LetterCodeMapping::getObservableItemId, LetterCodeMapping::getLetterCode);
```

bpasquer commented:
@nspool There is one problem. The template needs to provide an entry for “Survey not Done”, but “Survey not done” is not an observable item and therefore doesn't have a record in the method_species table.

nspool commented:
@bpasquer added the Survey Not Done row to each of the CSV export files.

jonescc commented on api/src/main/java/au/org/aodn/nrmn/restapi/repository/LetterCodeRespository.java:
```diff
@@ -0,0 +1,40 @@
+            + "FROM {h-schema}observable_item_ref oir JOIN stage7 s7 on s7.observable_item_id = oir.observable_item_id "
+            + "JOIN {h-schema}obs_item_type_ref oitr ON oitr.obs_item_type_id = oir.obs_item_type_id "
+            + "WHERE oitr.obs_item_type_id = ANY (ARRAY[1, 2]) ORDER BY letterCode", nativeQuery = true)
+    List<LetterCodeMapping> findByCriteria(@Param("siteIds") List<Integer> siteIds);
```
findByCriteria is used elsewhere for generic entity lookups during validation.   I'd give this one a different name reflecting its purpose so that this usage doesn't get confused with the other.

nspool commented:
@jonescc `SpeciesWithAttributes` and associated repository was used in the original species CSV PR. It was subsequently replaced by the `UiSpeciesAttributes` and the `ObservationRepository` but not in the original PR.

nspool commented:
Replaced the usage of SpeciesWithAttributes with UiSpeciesAttributes and also moved the letter code mapping to the template service. Also updated unit tests.

nspool commented:
Stack available here: http://nrmn-species-csv.dev.aodn.org.au/

nspool commented:
Redeploying stack as pointing at wrong branch 🤦‍♂️

jonescc reviewed (changes requested):
Looks good - but as discussed we're including extra species in species csvs that shouldn't be there.

e.g. for South Australia - we are getting invertebrates  included in the method 1 list

nspool commented:
@jonescc modified to use the `methods_species` table for the method ID check. Stack redeployed at  http://nrmn-species-csv.dev.aodn.org.au/.
